test/mtk.jl

```diff
@@ -32,8 +34,116 @@ prob = ODEProblem(sys, u0, tspan, p, jac = true)
 sol = solve(prob, Tsit5())
 mtkparams = SciMLSensitivity.parameter_values(sol)
 
-gt = rand(5501)
+gt = rand(1495)
 dmtk, = Zygote.gradient(mtkparams) do p
-    new_sol = solve(prob, Rosenbrock23(), p = p)
+    new_sol = solve(prob, Tsit5(), p = p)
+    Zygote.ChainRules.ChainRulesCore.ignore_derivatives() do
+        @test all(isapprox.(new_sol[x + y + z + 2 * β - w], 0, atol = 1e-12))
+    end
     mean(abs.(new_sol[sys.x] .- gt))
 end
+
+# Force DAE handling for Initialization
+
+eqs = [D(D(x)) ~ σ * (y - x),
+    D(y) ~ x * (ρ - z) - y,
+    D(z) ~ x * y - β * z,
+    w ~ x + y + z + 2 * β,
+    0 ~ x^2 + y^2 - w2^2
+]
+
+@mtkbuild sys = ODESystem(eqs, t)
+
+u0_incorrect = [D(x) => 2.0,
+    x => 1.0,
+    y => 0.0,
+    z => 0.0]
+
+p = [σ => 28.0,
+    ρ => 10.0,
+    β => 8 / 3]
+
+tspan = (0.0, 100.0)
+
+# Check that the gradients for the `solve` are the same both for an initialization
+# (for the algebraic variables) initialized poorly (therefore needs correction with BrownBasicInit)
+# and with the initialization corrected to satisfy the algebraic equation
+prob_incorrectu0 = ODEProblem(sys, u0_incorrect, tspan, p, jac = true, guesses = [w2 => 0.0])
+mtkparams_incorrectu0 = SciMLSensitivity.parameter_values(prob_incorrectu0)
+
+u0_timedep = [D(x) => 2.0,
+    x => 1.0,
+    y => t,
+    z => 0.0,
+    w2 => 0.0,]
+# this ensures that `y => t` is not applied in the adjoint equation
+# If the MTK init is called for the reverse, then `y0` in the backwards
+# pass will be extremely far off and cause an incorrect gradient
+prob_timedepu0 = ODEProblem(sys, u0_timedep, tspan, p, jac = true, guesses = [w2 => 0.0])
+mtkparams_timedepu0 = SciMLSensitivity.parameter_values(prob_incorrectu0)
+
+u0_correct = [D(x) => 2.0,
+    x => 1.0,
+    y => 0.0,
+    z => 0.0,
+    w2 => -1.0,]
+prob_correctu0 = ODEProblem(sys, u0_correct, tspan, p, jac = true, guesses = [w2 => -1.0])
+mtkparams_correctu0 = SciMLSensitivity.parameter_values(prob_correctu0)
+prob_correctu0.u0[5] = -1.0
```

Comment:
@AayushSabharwal note that this is the issue of not passing over the initial values and guess values to u0. Without this `CheckInit` will fail. It's a minor issue but it is something we should handle over time.

Reply:
Yeah I'm working on using the hook in DiffEqBase. It's easy enough to implement, but _somehow_ the returned problem's u0 isn't being respected